test/parallel/test-assert-objects.js

```diff
@@ -0,0 +1,399 @@
+  });
+
+  it('should not strictly compare two objects with different Error instances', function() {
+    const obj1 = { error: new Error('Test error 1') };
+    const obj2 = { error: new Error('Test error 2') };
+    assert.throws(() => assert.matchObjectStrict(obj1, obj2), Error);
+  });
+
+  it('should strictly compare two objects with the same TypedArray instance', function() {
+    const typedArray = new Uint8Array([1, 2, 3]);
+    const obj1 = { typedArray: typedArray };
+    const obj2 = { typedArray: typedArray };
+    assert.matchObjectStrict(obj1, obj2);
+  });
+
+  it('should not strictly compare two objects with different TypedArray instances and content', function() {
+    const obj1 = { typedArray: new Uint8Array([1, 2, 3]) };
+    const obj2 = { typedArray: new Uint8Array([4, 5, 6]) };
+    assert.throws(() => assert.matchObjectStrict(obj1, obj2), Error);
+  });
+
+  it('should not strictly compare two objects with different TypedArray instances', function() {
+    const obj1 = { typedArray: new Uint8Array([1, 2, 3]) };
+    const obj2 = { typedArray: new Uint8Array([1, 2, 3]) };
+    assert.matchObjectStrict(obj1, obj2);
+  });
+
+  it('should strictly compare two Map objects with identical entries', function() {
+    const map1 = new Map([
+      ['key1', 'value1'],
+      ['key2', 'value2'],
+    ]);
+    const map2 = new Map([
+      ['key1', 'value1'],
+      ['key2', 'value2'],
+    ]);
+    assert.matchObjectStrict(map1, map2);
+  });
+
+  it('should not strictly compare two Map objects with different entries', function() {
+    const map1 = new Map([
+      ['key1', 'value1'],
+      ['key2', 'value2'],
+    ]);
+    const map2 = new Map([
+      ['key1', 'value1'],
+      ['key3', 'value3'],
+    ]);
+    assert.throws(() => assert.matchObjectStrict(map1, map2), Error);
+  });
+
+  it('should strictly compare two Set objects with identical values', function() {
+    const set1 = new Set(['value1', 'value2']);
+    const set2 = new Set(['value1', 'value2']);
+    assert.matchObjectStrict(set1, set2);
+  });
+
+  it('should strictly compare two Map objects from different realms with identical entries', function() {
+    const map1 = new vm.runInNewContext('new Map([["key1", "value1"], ["key2", "value2"]])');
+    const map2 = new Map([
+      ['key1', 'value1'],
+      ['key2', 'value2'],
+    ]);
+    assert.matchObjectStrict(map1, map2);
+  });
+
+  it('should not strictly compare two Set objects from different realms with different values', function() {
+    const set1 = new vm.runInNewContext('new Set(["value1", "value2"])');
+    const set2 = new Set(['value1', 'value3']);
+    assert.throws(() => assert.matchObjectStrict(set1, set2), Error);
+  });
+
+  it('should not strictly compare two Set objects with different values', function() {
+    const set1 = new Set(['value1', 'value2']);
+    const set2 = new Set(['value1', 'value3']);
+    assert.throws(() => assert.matchObjectStrict(set1, set2), Error);
+  });
```

Comment:
```suggestion
  });
  
  it('should compare plain objects from different realms', function() {
    assert.matchObjectStrict(
      vm.runInNewContext('{ a: 1, b: 2n, c: '3', d: /4/, e: new Set([5]), f: [6], g: new Uint8Array }'),
      { d: /4/, e: new Set([5]), f: [6], g: new Uint8Array() }
    );
  });
```

Reply:
This will throw. While objects, maps, and sets will recurse and rely on `deepEqual` for the values arrays are immediately compared using `deepEqual` which treats arrays from different realms as not equal